packages/liveblocks-core/src/connection.ts

```diff
@@ -0,0 +1,756 @@
+    const newPublicStatus = toPublicConnectionStatus(newState);
+    statusDidChange.notify(newPublicStatus);
+
+    if (oldPublicStatus === "open" && newPublicStatus !== "open") {
+      didDisconnect.notify();
+    } else if (oldPublicStatus !== "open" && newPublicStatus === "open") {
+      didConnect.notify();
+    }
+    oldPublicStatus = newPublicStatus;
+  });
+
+  return {
+    statusDidChange: statusDidChange.observable,
+    didConnect: didConnect.observable,
+    didDisconnect: didDisconnect.observable,
+  };
+}
+
+function createStateMachine<T extends BaseAuthResult>(delegates: Delegates<T>) {
+  // Create observable event sources, which this machine will call into when
+  // specific events happen
+  const onMessage = makeEventSource<IWebSocketMessageEvent>();
+
+  // Emitted whenever the server deliberately closes the connection for
+  // a specific Liveblocks reason
+  const onLiveblocksError = makeEventSource<LiveblocksError>();
+
+  const initialContext: Context & { token: T | null } = {
+    token: null,
+    socket: null,
+
+    // Bumped to the next "tier" every time a connection attempt fails (no matter
+    // whether this is for the authentication server or the websocket server).
+    // Reset every time a connection succeeded.
+    backoffDelay: LOW_DELAY,
+    // numRetries: 0,
+  };
+
+  const fsm = new FSM<Context, Event, State>(initialContext)
+    .addState("@idle.initial")
+    .addState("@idle.failed")
+    .addState("@auth.busy")
+    .addState("@auth.backoff")
+    .addState("@connecting.busy")
+    .addState("@connecting.backoff")
+    .addState("@ok.connected")
+    .addState("@ok.awaiting-pong");
+
+  //
+  // Configure events that can happen from anywhere
+  //
+  // It's always possible to explicitly get a .reconnect() or .disconnect()
+  // from the user.
+  //
+  fsm.addTransitions("*", {
+    RECONNECT: {
+      target: "@auth.backoff",
+      assign: increaseBackoffDelay,
+    },
+
+    DISCONNECT: "@idle.initial",
+  });
+
+  //
+  // Configure the @idle.* states
+  //
+  fsm.addTransitions("@idle.*", {
+    CONNECT: (_, ctx) =>
+      // If we still have a known token, try to reconnect to the socket directly,
+      // otherwise, try to obtain a new token
+      ctx.token !== null ? "@connecting.busy" : "@auth.busy",
+  });
+
+  //
+  // Configure the @auth.* states
+  //
+  fsm
+    .addTransitions("@auth.backoff", {
+      NAVIGATOR_ONLINE: {
+        target: "@auth.busy",
+        assign: { backoffDelay: LOW_DELAY },
+      },
+    })
+    .addTimedTransition(
+      "@auth.backoff",
+      (ctx) => ctx.backoffDelay,
+      "@auth.busy"
+    )
+
+    .onEnterAsync(
+      "@auth.busy",
+
+      () =>
+        Promise.race([delegates.authenticate(), timeoutAfter(AUTH_TIMEOUT)]),
+
+      // On successful authentication
+      (okEvent) => ({
+        target: "@connecting.busy",
+        assign: {
+          token: okEvent.data as BaseAuthResult,
+          backoffDelay: LOW_DELAY,
```

Comment:
this is just resetting it back to the quicker backoff in case of future failures right? We're not actually doing any delays or anything here on success?

Reply:
Correct, it's a reset indeed, setting it back to the lowest delay (it's never 0). In a prior iteration, I named that `LOW_DELAY` variable `RESET_DELAY`, which may make this easier. I'll change that back, since it would have been clearer. Fixed in 7117ea4066ee366487336dd735dcd171986515b7.